### DISCUSSION TITLE
v7.2.0-preview.5 Release of PowerShell

### DISCUSSION THREAD

**Discussion by adityapatwardhan** [PowerShell/PowerShell#15236] [General]
### Breaking Changes

- Make PowerShell Linux deb and RPM packages universal (#15109)
- Enforce AppLocker Deny configuration before Execution Policy Bypass configuration (#15035)
- Disallow mixed dash and slash in command line parameter prefix (#15142) (Thanks @davidBar-On!)

### Experimental Features

- `PSNativeCommandArgumentPassing`: Use `ArgumentList` for native executable invocation (breaking change) (#14692)

### Engine Updates and Fixes

- Add `IArgumentCompleterFactory` for parameterized `ArgumentCompleters` (#12605) (Thanks @powercode!)

### General Cmdlet Updates and Fixes

- Fix SSH remoting connection never finishing with misconfigured endpoint (#15175)
- Respect `TERM` and `NO_COLOR` environment variables for `$PSStyle` rendering (#14969)
- Use `ProgressView.Classic` when Virtual Terminal is not supported (#15048)
- Fix `Get-Counter` issue with `-Computer` parameter (#15166) (Thanks @krishnayalavarthi!)
- Fix redundant iteration while splitting lines (#14851) (Thanks @hez2010!)
- Enhance `Remove-Item -Recurse` to work with OneDrive (#14902) (Thanks @iSazonov!)
- Change minimum depth to 0 for `ConvertTo-Json` (#14830) (Thanks @kvprasoon!)
- Allow `Set-Clipboard` to accept empty string (#14579)
- Turn on and off `DECCKM` to modify keyboard mode for Unix native commands to work correctly (#14943)
- Fall back to `CopyAndDelete()` when `MoveTo()` fails due to an `IOException` (#15077)

### Code Cleanup

<details>

<summary>

<p>We thank the following contributors!</p>
<p>@xtqqczze, @iSazonov, @ZhiZe-ZG</p>

</summary>

<ul>
<li>Update .NET to <code>6.0.0-preview.3</code> (#15221)</li>
<li>Add space before comma to hosting test to fix error reported by <code>SA1001</code> (#15224)</li>
<li>Add <code>SecureStringHelper.FromPlainTextString</code> helper method for efficient secure string creation (#14124) (Thanks @xtqqczze!)</li>
<li>Use static lambda keyword (#15154) (Thanks @iSazonov!)</li>
<li>Remove unnecessary <code>Array</code> -&gt; <code>List</code> -&gt; <code>Array</code> conversion in <code>ProcessBaseCommand.AllProcesses</code> (#15052) (Thanks @xtqqczze!)</li>
<li>Standardize grammar comments in Parser.cs (#15114) (Thanks @ZhiZe-ZG!)</li>
<li>Enable <code>SA1001</code>: Commas should be spaced correctly (#14171) (Thanks @xtqqczze!)</li>
<li>Refactor <code>MultipleServiceCommandBase.AllServices</code> (#15053) (Thanks @xtqqczze!)</li>
</ul>

</details>

### Tools

- Use Unix line endings for shell scripts (#15180) (Thanks @xtqqczze!)

### Tests

- Add the missing tag in Host Utilities tests (#14983)
- Update `copy-props` version in `package.json` (#15124)

### Build and Packaging Improvements

<details>

<summary>

<p>We thank the following contributors!</p>
<p>@JustinGrote</p>

</summary>

<ul>
<li>Fix <code>yarn-lock</code> for <code>copy-props</code> (#15225)</li>
<li>Make package validation regex accept universal Linux packages (#15226)</li>
<li>Bump NJsonSchema from 10.4.0 to 10.4.1 (#15190)</li>
<li>Make MSI and EXE signing always copy to fix daily build (#15191)</li>
<li>Sign internals of EXE package so that it works correctly when signed (#15132)</li>
<li>Bump Microsoft.NET.Test.Sdk from 16.9.1 to 16.9.4 (#15141)</li>
<li>Update daily release tag format to  work with new Microsoft Update work (#15164)</li>
<li>Feature: Add Ubuntu 20.04 Support to install-powershell.sh (#15095) (Thanks @JustinGrote!)</li>
<li>Treat rebuild branches like release branches (#15099)</li>
<li>Update WiX to 3.11.2 (#15097)</li>
<li>Bump NJsonSchema from 10.3.11 to 10.4.0 (#15092)</li>
<li>Allow patching of preview releases (#15074)</li>
<li>Bump Newtonsoft.Json from 12.0.3 to 13.0.1 (#15084, #15085)</li>
<li>Update the <code>minSize</code> build package filter to be explicit (#15055)</li>
<li>Bump NJsonSchema from 10.3.10 to 10.3.11 (#14965)</li>
</ul>

</details>

### Documentation and Help Content

- Merge `7.2.0-preview.4` changes to master (#15056)
- Update `README` and `metadata.json` (#15046)
- Fix broken links for `dotnet` CLI (#14937)

[7.2.0-preview.5]: https://github.com/PowerShell/PowerShell/compare/v7.2.0-preview.4...v7.2.0-preview.5

### SHA256 Hashes of the release artifacts

- powershell-7.2.0-preview.5-linux-alpine-x64.tar.gz
  - 823E644B706BA4D8CF3F177F19B77012BF0789B2A5861E66FA51CE120C117281
- powershell-7.2.0-preview.5-linux-arm32.tar.gz
  - 7268C29913C9BEE407F3F1D80A4D82F3961497B6AFC992006509219F5A4169FF
- powershell-7.2.0-preview.5-linux-arm64.tar.gz
  - A60DCE3F6AB109A71BEF7421504E47934C7801D8538ED153E0A1DD43F806EBD4
- powershell-7.2.0-preview.5-linux-x64.tar.gz
  - 9EF2746989096226789C101A93789950973E3098D10BC05EE56569B00DE2BB5A
- powershell-7.2.0-preview.5-linux-x64-fxdependent.tar.gz
  - 0001DB9B1F994484C6E6E9457FDCB0D7735065FFA8E6ECCEC44207C37CD16518
- powershell-7.2.0-preview.5-osx-x64.pkg
  - D4464E9F8441B021CFE4A0BF15B7EDBBF047ECE3534F45CE85FD1AFB95C672E0
- powershell-7.2.0-preview.5-osx-x64.tar.gz
  - 3648611C63070BB51A6B06EEB75F2C4C525A0571FA9D62394FEAE52A08F3CC04
- PowerShell-7.2.0-preview.5-win-arm32.zip
  - C6901D20CA032B6C14B491F27E7E10CCB6F52A1137FF35D4624532F7D68FDE59
- PowerShell-7.2.0-preview.5-win-arm64.zip
  - 0A877120AA3B38FBBFA7670CA456C1D43714A4675EA2D8C54BCA63BB9739D87D
- PowerShell-7.2.0-preview.5-win-fxdependent.zip
  - 98299865345E1DE504494F7EBE22A5D6B9037469F8FDE63ADF73FD079EB3DAFA
- PowerShell-7.2.0-preview.5-win-fxdependentWinDesktop.zip
  - 471B6D2967A80610907E016CA8C903CC661BDBB09B459C1FABD541FFEBAD1A7A
- PowerShell-7.2.0-preview.5-win-x64.msi
  - 2D89D41B4DAE52B643F8ECB0316C954E437D516F9894A3B146F704F603EAE985
- PowerShell-7.2.0-preview.5-win-x64.zip
  - BF3E8343DC75D79B1FFEFB488A5BEB36596612A05EDEA68A7F4173D863BAE3EE
- PowerShell-7.2.0-preview.5-win-x86.msi
  - 0E9C62631D65A66E4F735BB166FEEE266140EA04AD8C24374125B7041DE5A8A3
- PowerShell-7.2.0-preview.5-win-x86.zip
  - 56CDE708F13A189BD49290DE75E527D71189A0EA284BFE37823BE2D48FE36DF2
- powershell-preview_7.2.0-preview.5-1.deb_amd64.deb
  - 7BAA7D3E12053844FC4B069302B00346D11E83B7BB59A25702C58B555C643705
- powershell-preview-7.2.0_preview.5-1.rh.x86_64.rpm
  - 95E2ECE9EC2BCE7F689E4E70821BFD2903E368F9C93F6B2A35D1A7B40168894D

<hr /><em>This discussion was created from the release <a href='https://github.com/PowerShell/PowerShell/releases/tag/v7.2.0-preview.5'>v7.2.0-preview.5 Release of PowerShell</a>.</em>

**Comment by adityapatwardhan** [comment]
Having some issues publishing Linux packages to package repositories. We are having a look.

**Reply by adityapatwardhan** [reply]
The packages have been published. Let us know if you see any issues.

**Comment by mklement0** [comment]
I'd like to draw attention to an **important new [experimental feature](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_Experimental_Features)** that is part of this release, **`PSNativeCommandArgumentPassing`: it aims to fix the longstanding problem with quoting when calling native executables** (see #1995).

With `PSNativeCommandArgumentPassing` in effect (which it true by default in _preview_ releases, not release candidates and regular releases):

* Existing workarounds for #1995 will break, unless you set new preference variable `$PSNativeCommandArgumentPassing` to `'Legacy'`

* Conversely, with value `'Standard'` (the default, at least in the context of the experimental feature), passing arguments to external programs:

  * now _fully_ works as it always should have on _Unix_-like platforms.

  * only _partially_ works as it always should have on Windows, because vital accommodations for high-profile CLIs, are missing - see #15143.

  * _breaks_ `cmd /c` (`cmd /k`) calls, where a command line for `cmd.exe` is passed as a _single string_ - which is a common workaround for the problems with the previous behavior: see #15239

**This experimental feature, if fully implemented - with accommodations on Windows and bugs fixed - has the potential to make quoting headaches when calling native programs go away once and for all (100% on Unix, 99.9...% on Windows), so it is important that it be exercised and put through its paces**, so that it can become a regular feature - whether by opt-_in_ or opt-_out_ is TBD (see https://github.com/PowerShell/PowerShell/issues/1995#issuecomment-820541951)

**Reply by mklement0** [reply]
Unfortunately, the experimental feature also breaks `--%` (the stop-parsing symbol): see #15261

**Reply by iSazonov** [reply]
The silly situation persists - it doesn't work well as a shell 😄

**Reply by mklement0** [reply]
Yes, unfortunately. I've just tried to summarize what's currently broken and also what vital accommodations are missing in an easier-to-grasp form [here](https://github.com/PowerShell/PowerShell/issues/15250#issuecomment-822030173) - after making the same points repeatedly, I hope it will finally get the attention that it needs.

**Reply by mklement0** [reply]
Unfortunately, the experimental feature also breaks passing arguments such as `-F:` (looks like a PowerShell parameter, _ends in `:`_) - see #15276

I'll stop reporting these problems piecemeal here, and will try to maintain a current list in https://github.com/PowerShell/PowerShell/issues/15143#issuecomment-822805995
